File: code/typescript-advanced/compare/l01_typed_keys.cs
// compare/l01_typed_keys.cs
// C# has no literal types: a string can't carry its payload's type, so a typed key object does
var hub = new Hub();
hub.On(HubEvents.NavigateToPlanet, data => Console.WriteLine($"navigate to {data.Target}"));
hub.On(HubEvents.Connected, data => Console.WriteLine($"{data.Connections} clients connected"));
hub.Receive("NavigateToPlanet", new NavigateToPlanet("saturn"));
hub.Receive("Connected", new Connected(2));

record NavigateToPlanet(string Target);
record Connected(int Connections);

// The key: a name for the wire, and a type parameter for the compiler
sealed record HubEvent<T>(string Name);

static class HubEvents
{
    public static readonly HubEvent<NavigateToPlanet> NavigateToPlanet = new("NavigateToPlanet");
    public static readonly HubEvent<Connected> Connected = new("Connected");
}

class Hub
{
    private readonly Dictionary<string, Action<object>> handlers = [];

    // T is inferred from the key, as TypeScript infers K from the string
    public void On<T>(HubEvent<T> hubEvent, Action<T> handler) => handlers[hubEvent.Name] = data => handler((T)data);

    public void Receive(string name, object data) => handlers[name](data);
}

File: code/typescript-advanced/compare/L01TypedKeys.java
// compare/L01TypedKeys.java
// Java has no literal types either: the typed key carries the payload type
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class L01TypedKeys {
    record NavigateToPlanet(String target) {}
    record Connected(int connections) {}

    record HubEvent<T>(String name, Class<T> type) {}

    static final HubEvent<NavigateToPlanet> NAVIGATE_TO_PLANET = new HubEvent<>("NavigateToPlanet", NavigateToPlanet.class);
    static final HubEvent<Connected> CONNECTED = new HubEvent<>("Connected", Connected.class);

    static final Map<String, Consumer<Object>> handlers = new HashMap<>();

    // The Class<T> in the key also allows a checked cast, which Java's erased T can't do on its own
    static <T> void on(HubEvent<T> event, Consumer<T> handler) {
        handlers.put(event.name(), data -> handler.accept(event.type().cast(data)));
    }

    public static void main(String[] args) {
        on(NAVIGATE_TO_PLANET, data -> System.out.println("navigate to " + data.target()));
        on(CONNECTED, data -> System.out.println(data.connections() + " clients connected"));
        handlers.get("NavigateToPlanet").accept(new NavigateToPlanet("saturn"));
        handlers.get("Connected").accept(new Connected(2));
        try {
            handlers.get("Connected").accept(new NavigateToPlanet("mars"));
        } catch (ClassCastException e) {
            System.out.println("ClassCastException: " + e.getMessage());
        }
    }
}
